### PR TITLE
update for frozen string literals

### PR DESCRIPTION
**Pull request by twalpole**
Frozen string literals - magic comment in all files and changes for tests to pass

### CODE REVIEW

**Comment by kgrz** [comment]
> ...for tests to pass

Are these test related to the Sinatra repo? any situation where the failure can be reproduced? The master branch looks green.

**Comment by twalpole** [comment]
Yes, it was Sinatra tests that were failing when using default frozen string literals - specifically attempting to concat to frozen strings, see the changes in base.rb.  They don't fail in master because it's not defaulting to frozen string literals.

File: lib/sinatra.rb [hunk 1]
```diff
@@ -1,3 +1,4 @@
+# frozen_string_literal: true
```
Comment (zzak) [inline lib/sinatra.rb]: This isn't needed for the one string allocated here "sinatra/main" is an acceptable allocation that I doubt will be repeated enough to affect performance

**Comment by zzak** [comment]
I also don't want to litter the test files with this comment, as it's only necessary for runtime performance

File: lib/sinatra/base.rb [hunk 2]
```diff
@@ -842,6 +842,7 @@ def compile_template(engine, data, options, views)
           body, path, line = settings.templates[data]
           if body
             body = body.call if body.respond_to?(:call)
+            body = body.dup if body.frozen?
```
Comment (zzak) [inline lib/sinatra/base.rb]: Can you provide a case where body is actually immutable from here? Maybe an empty response?

File: lib/sinatra/base.rb [hunk 3]
```diff
@@ -857,7 +858,12 @@ def compile_template(engine, data, options, views)
             template.new(path, 1, options)
           end
         when Proc, String
-          body = data.is_a?(String) ? Proc.new { data } : data
+          body = if data.is_a?(String)
```
Comment (zzak) [inline lib/sinatra/base.rb]: same here

File: lib/sinatra/base.rb [hunk 4]
```diff
@@ -348,7 +348,7 @@ def attachment(filename = nil, disposition = :attachment)
       response['Content-Disposition'] = disposition.to_s
       if filename
         params = '; filename="%s"' % File.basename(filename)
-        response['Content-Disposition'] << params
+        response['Content-Disposition'] += params
```
Comment (zzak) [inline lib/sinatra/base.rb]: Here as well, I'm not sure this change is necessary as we haven't seen any failures in this case

Comment (aselder) [inline lib/sinatra/base.rb]: @zzak I just ran across this in one of our projects. See https://github.com/sinatra/sinatra/pull/1137. We tried to freeze the constant we were using to set dispositions and got lots of errors..

Also this fixed a possible unintended side effect bug. If you hold a reference to the string you pass as the disposition, you'll find that string changed after the call.

```
x = 'inline'
send_file(some_path, disposition: x)
x == 'inline' #=> false
```

Comment (zzak) [inline lib/sinatra/base.rb]: @aselder Thanks for your investigation!

I will consider your patch, and this one but unfortunately I don't want to merge all of these comments in the test files.

**Pull request closed without merging**